File: packages/python/secrefs/resolver.py
from __future__ import annotations

import asyncio
import os
from dataclasses import dataclass
from typing import Dict, List, Mapping, Optional, Tuple

from .parser import ParsedSecretRef, is_secret_ref, parse_secret_ref
from .providers.base import SecretFetchRequest, SecretProvider

ProviderRegistry = Mapping[str, SecretProvider]


@dataclass(frozen=True)
class ResolutionFailure:
    key: str
    ref: str
    message: str


class SecRefsResolutionError(Exception):
    def __init__(self, errors: List[ResolutionFailure]) -> None:
        self.errors = errors
        lines = "\n".join(f"  - {e.key}: {e.ref} -> {e.message}" for e in errors)
        super().__init__(f"Failed to resolve {len(errors)} secret reference(s):\n{lines}")


@dataclass(frozen=True)
class CheckResult:
    key: str
    ref: str
    provider: str
    ok: bool
    message: Optional[str] = None  # present only when ok is False; never a secret value


async def _resolve_one(ref: ParsedSecretRef, providers: ProviderRegistry) -> str:
    provider = providers.get(ref.provider)
    if provider is None:
        available = ", ".join(providers.keys()) or "none configured"
        raise ValueError(f'unknown provider "{ref.provider}" (available: {available})')
    return await provider.fetch_one(SecretFetchRequest(path=ref.path, field=ref.field))


async def expand_key_value_map(
    input_map: Mapping[str, Optional[str]],
    providers: ProviderRegistry,
    *,
    strict: bool = True,
) -> Dict[str, str]:
    """
    Expands every `sec://` value in a plain key/value map, resolving all
    references concurrently. Non-reference values pass through untouched.
    Raises SecRefsResolutionError aggregating every failed reference.
    """
    output: Dict[str, str] = {}
    pending: List[Tuple[str, ParsedSecretRef]] = []

    for key, value in input_map.items():
        if value is None:
            continue
        if not is_secret_ref(value):
            output[key] = value
            continue
        try:
            pending.append((key, parse_secret_ref(value)))
        except ValueError:
            if strict:
                raise
            output[key] = value

    if not pending:
        return output

    results = await asyncio.gather(
        *(_resolve_one(ref, providers) for _, ref in pending), return_exceptions=True
    )

    errors: List[ResolutionFailure] = []
    for (key, ref), result in zip(pending, results):
        if isinstance(result, BaseException):
            errors.append(ResolutionFailure(key=key, ref=ref.raw, message=str(result)))
        else:
            output[key] = result

    if errors:
        raise SecRefsResolutionError(errors)

    return output


async def expand_environ(providers: ProviderRegistry, *, strict: bool = True) -> List[str]:
    """Expands sec:// values found in os.environ, mutating it in place."""
    resolved = await expand_key_value_map(dict(os.environ), providers, strict=strict)
    changed_keys: List[str] = []
    for key, value in resolved.items():
        if os.environ.get(key) != value:
            os.environ[key] = value
            changed_keys.append(key)
    return changed_keys


async def check_references(
    input_map: Mapping[str, Optional[str]], providers: ProviderRegistry
) -> List[CheckResult]:
    """
    Dry-run validation: resolves every sec:// reference in `input_map` but
    reports only ok/failure - secret values are never returned or logged.
    """
    results: List[CheckResult] = []
    parsed_entries: List[Tuple[str, ParsedSecretRef]] = []

    for key, value in input_map.items():
        if not isinstance(value, str) or not is_secret_ref(value):
            continue
        try:
            parsed_entries.append((key, parse_secret_ref(value)))
        except ValueError as exc:
            results.append(
                CheckResult(key=key, ref=value, provider="unknown", ok=False, message=str(exc))
            )

    outcomes = await asyncio.gather(
        *(_resolve_one(ref, providers) for _, ref in parsed_entries), return_exceptions=True
    )

    for (key, ref), outcome in zip(parsed_entries, outcomes):
        results.append(
            CheckResult(
                key=key,
                ref=ref.raw,
                provider=ref.provider,
                ok=not isinstance(outcome, BaseException),
                message=str(outcome) if isinstance(outcome, BaseException) else None,
            )
        )

    return results

File: packages/node/src/resolver.ts
import { type ParsedSecretRef, isSecretRef, parseSecretRef } from "./parser.js";
import {
  errorMessage,
  SecretFetchError,
  type ISecretProvider,
  type SecretFetchRequest,
} from "./providers/base.js";
import type { SecretErrorKind } from "./providers/errors.js";

export type ProviderRegistry = Record<string, ISecretProvider>;

export interface ExpandOptions {
  providers: ProviderRegistry;
  /**
   * When true (default), a value that starts with `sec://` but fails to
   * parse throws immediately. When false, such values are left untouched.
   * This only affects syntactically malformed references - unknown
   * providers and provider-side fetch failures always surface as errors,
   * aggregated in a {@link SecRefsResolutionError}.
   */
  strict?: boolean;
}

export interface ResolutionFailure {
  /** The env var / map key the reference was assigned to. */
  key: string;
  /** The original `sec://` string. */
  ref: string;
  message: string;
  /** Whose problem this is. `undefined` when the failure came from
   * somewhere that doesn't classify (a malformed reference, say). */
  kind?: SecretErrorKind;
  /** Provider alias the reference named, for grouping auth failures. */
  provider?: string;
  /** What to run to fix it, for auth failures. */
  remedy?: string;
}

/**
 * Renders failures the way they should be read.
 *
 * One expired credential produces one failure per reference, and listing
 * them individually - `DB_PASSWORD: could not fetch...`, four times -
 * reads as four broken secrets and sends people to check their secrets.
 * Auth failures are therefore collapsed per provider, stated as an
 * environment problem, and given the command that fixes them. Path
 * failures stay itemised, because there the reference really is the
 * thing at fault.
 */
function formatFailures(errors: ResolutionFailure[]): string {
  const auth = errors.filter((e) => e.kind === "auth");
  const rest = errors.filter((e) => e.kind !== "auth");
  const lines: string[] = [];

  for (const provider of [...new Set(auth.map((e) => e.provider ?? "unknown"))]) {
    const group = auth.filter((e) => (e.provider ?? "unknown") === provider);
    lines.push(`Cannot authenticate to provider "${provider}".`);
    lines.push(`  ${group[0]!.message}`);
    const remedy = group.find((e) => e.remedy)?.remedy;
    if (remedy) lines.push(`  ${remedy}`);
    lines.push(`  Not resolved: ${group.map((e) => e.key).join(", ")}`);
  }

  if (rest.length > 0) {
    if (lines.length > 0) lines.push("");
    lines.push(`Failed to resolve ${rest.length} secret reference(s):`);
    for (const e of rest) lines.push(`  - ${e.key}: ${e.ref} -> ${e.message}`);
  }

  return lines.join("\n");
}

export class SecRefsResolutionError extends Error {
  constructor(public readonly errors: ResolutionFailure[]) {
    super(formatFailures(errors));
    this.name = "SecRefsResolutionError";
  }

  /** True when every failure was an environment/auth problem, so a caller
   * can tell "your credentials lapsed" from "your references are wrong". */
  get isAuthOnly(): boolean {
    return this.errors.length > 0 && this.errors.every((e) => e.kind === "auth");
  }
}

export interface CheckResult {
  key: string;
  ref: string;
  provider: string;
  ok: boolean;
  /** Present only when ok is false. Never contains the secret value. */
  message?: string;
}

async function resolveOne(
  ref: ParsedSecretRef,
  providers: ProviderRegistry,
): Promise<string> {
  const provider = providers[ref.provider];
  if (!provider) {
    const available = Object.keys(providers).join(", ") || "none configured";
    throw new Error(`unknown provider "${ref.provider}" (available: ${available})`);
  }
  const request: SecretFetchRequest = { path: ref.path, field: ref.field };
  return provider.fetchOne(request);
}

/**
 * Expands every `sec://` value in a plain key/value map, resolving all
 * references concurrently via `Promise.allSettled`. Non-reference values
 * pass through untouched. Never writes anything to disk - the caller
 * decides what to do with the returned map (assign to `process.env`,
 * template into a string, etc).
 *
 * Throws {@link SecRefsResolutionError} aggregating every failed
 * reference if any fail to resolve.
 */
export async function expandKeyValueMap(
  input: Record<string, string | undefined>,
  options: ExpandOptions,
): Promise<Record<string, string>> {
  const strict = options.strict ?? true;
  const output: Record<string, string> = {};
  const pending: { key: string; ref: ParsedSecretRef }[] = [];

  for (const [key, value] of Object.entries(input)) {
    if (value === undefined) continue;
    if (!isSecretRef(value)) {
      output[key] = value;
      continue;
    }
    try {
      pending.push({ key, ref: parseSecretRef(value) });
    } catch (err) {
      if (strict) throw err;
      output[key] = value;
    }
  }

  if (pending.length === 0) {
    return output;
  }

  const settled = await Promise.allSettled(
    pending.map(({ ref }) => resolveOne(ref, options.providers)),
  );

  const errors: ResolutionFailure[] = [];
  settled.forEach((result, i) => {
    const { key, ref } = pending[i] as { key: string; ref: ParsedSecretRef };
    if (result.status === "fulfilled") {
      output[key] = result.value;
    } else {
      const reason = result.reason;
      errors.push({
        key,
        ref: ref.raw,
        message: errorMessage(reason),
        kind: reason instanceof SecretFetchError ? reason.kind : undefined,
        provider: reason instanceof SecretFetchError ? reason.provider : ref.provider,
        remedy: reason instanceof SecretFetchError ? reason.remedy : undefined,
      });
    }
  });

  if (errors.length > 0) {
    throw new SecRefsResolutionError(errors);
  }

  return output;
}

/**
 * Expands `sec://` values found in `process.env`, mutating it in place.
 * Returns the list of env var names that were rewritten.
 */
export async function expandProcessEnv(options: ExpandOptions): Promise<string[]> {
  const resolved = await expandKeyValueMap(process.env, options);
  const changedKeys: string[] = [];
  for (const [key, value] of Object.entries(resolved)) {
    if (process.env[key] !== value) {
      process.env[key] = value;
      changedKeys.push(key);
    }
  }
  return changedKeys;
}

/**
 * Dry-run validation: resolves every `sec://` reference found in `input`
 * but reports only ok/failure per reference - the secret values themselves
 * are never returned or logged. Used by `secrefs check`.
 */
export async function checkReferences(
  input: Record<string, string | undefined>,
  options: ExpandOptions,
): Promise<CheckResult[]> {
  const results: CheckResult[] = [];
  const parsedEntries: { key: string; ref: ParsedSecretRef }[] = [];

  for (const [key, value] of Object.entries(input)) {
    if (value === undefined || !isSecretRef(value)) continue;
    try {
      parsedEntries.push({ key, ref: parseSecretRef(value) });
    } catch (err) {
      results.push({ key, ref: value, provider: "unknown", ok: false, message: errorMessage(err) });
    }
  }

  const settled = await Promise.allSettled(
    parsedEntries.map(({ ref }) => resolveOne(ref, options.providers)),
  );

  settled.forEach((result, i) => {
    const { key, ref } = parsedEntries[i] as { key: string; ref: ParsedSecretRef };
    results.push({
      key,
      ref: ref.raw,
      provider: ref.provider,
      ok: result.status === "fulfilled",
      message: result.status === "rejected" ? errorMessage(result.reason) : undefined,
    });
  });

  return results;
}
